From: Zhu Yanhai <gaoyang.zyh@taobao.com>
Date: Wed, 17 Apr 2013 11:48:28 +0800
Subject: [PATCH] net: core: refuse to create new network devices except lo in non-init namespace
Patch-mainline: No. This is a workaround, upstream has perfect fix.
References: http://bugfree.corp.taobao.com/bug/226888

Creating new network devices in non-init namespace in the current kernel would make
the kernel panic. Generally it's because a not fully initialized kobject got used
to create some sub-dirs or attrtibutes underneath it.

The upstream 3.8+ kernel can perfectly support such opearations with sysfs-namespace-aware
patchset, therefore it's desirable to backport them if we have such demand someday.

See http://bugfree.corp.taobao.com/bug/226888 for more.

Signed-off-by: Zhu Yanhai <gaoyang.zyh@taobao.com>

diff --git a/net/core/dev.c b/net/core/dev.c
index 17cd90c..617b361 100644

Acked-by: 

--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -5364,6 +5364,11 @@ int register_netdevice(struct net_device *dev)
 	BUG_ON(dev->reg_state != NETREG_UNINITIALIZED);
 	BUG_ON(!net);
 
+	if (net != &init_net && strcmp(dev->name, "lo") != 0) {
+		ret = -EPERM;
+		goto out;
+	}
+
 	spin_lock_init(&dev->addr_list_lock);
 	netdev_set_addr_lockdep_class(dev);
 	netdev_init_queue_locks(dev);
-- 
1.7.4.1

